src/cliffordsynthesis/encoding/MultiGateEncoder.cpp

```diff
@@ -132,19 +134,61 @@ void MultiGateEncoder::assertTwoQubitGateOrderConstraints(
   }
 
   // gate variables of the current and the next time step
-  const auto& current = vars.gC[pos][ctrl][trgt];
-  const auto& gSNow   = vars.gS[pos];
-  const auto& gCNext  = vars.gC[pos + 1];
+  const auto& gSNow  = vars.gS[pos];
+  const auto& gCNext = vars.gC[pos + 1];
 
   // two identical CNOTs may not be applied in a row because they would cancel.
-  lb->assertFormula(LogicTerm::implies(current, !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gCNext[trgt][ctrl]));
+
+  // no gate on both qubits => no CNOT on them in the next time step.
+  constexpr auto noneIndex = gateToIndex(qc::OpType::None);
+  const auto     noGate    = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
+  // similarly,
+  // H - X   c - H
+  //     | = |
+  // H - c   X - H
+  // i.e., Hadamards on both qubits => no CNOT on them in the next time step.
+  constexpr auto hIndex = gateToIndex(qc::OpType::H);
+  const auto     hh     = gSNow[hIndex][ctrl] && gSNow[hIndex][trgt];
 
-  // if no gate is applied to both qubits, no CNOT on them can be applied in the
-  // next time step.
-  const auto noneIndex     = gateToIndex(qc::OpType::None);
-  const auto noGate        = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
   const auto noFurtherCnot = !gCNext[ctrl][trgt] && !gCNext[trgt][ctrl];
-  lb->assertFormula(LogicTerm::implies(noGate, noFurtherCnot));
+  lb->assertFormula(LogicTerm::implies(noGate || hh, noFurtherCnot));
```

Comment:
I really hate to bother you again with something. But the realization here still is not perfect.
The H constraints are now included twice.
The "proper" left-hand side for the implication would be `noGateOnBoth || ((X || Y || Z || H on control) && (X || Y || Z || H on target)`

Reply:
I even thought about it and forgot it again. Good catch